packager/VisibilityChecker.cc

```diff
@@ -0,0 +1,558 @@
+            it = std::adjacent_find(it, this->imports.end(), [](auto &l, auto &r) { return l.package == r.package; });
+            if (it == this->imports.end()) {
+                break;
+            }
+
+            // find the end of the region of duplicated imports
+            auto end =
+                std::find_if_not(it, this->imports.end(), [sym = it->package](auto &e) { return sym == e.package; });
+
+            auto first = it;
+
+            // Treat the first import as the authoritative one, and report errors for the rest.
+            std::advance(it, 1);
+
+            for (; it != end; ++it) {
+                if (auto e = ctx.beginError(it->importLoc, core::errors::Packager::InvalidImportOrExport)) {
+                    e.setHeader("Duplicate package import `{}`", it->package.show(ctx));
+                    e.addErrorLine(ctx.locAt(first->importLoc), "Previous package import found here");
+                    e.replaceWith("Remove import", ctx.locAt(it->importLoc), "");
+                }
+            }
+        }
+
+        this->imports.clear();
+    }
+
+    static std::vector<ast::ParsedFile> run(core::GlobalState &gs, WorkerPool &workers,
+                                            std::vector<ast::ParsedFile> files) {
+        Timer timeit(gs.tracer(), "packager.import_checker");
+        auto resultq = std::make_shared<BlockingBoundedQueue<ast::ParsedFile>>(files.size());
+        auto fileq = std::make_shared<ConcurrentBoundedQueue<ast::ParsedFile>>(files.size());
+
+        for (auto &file : files) {
+            fileq->push(std::move(file), 1);
+        }
+        files.clear();
+
+        workers.multiplexJob("ImportChecker", [&gs, fileq, resultq]() {
+            ast::ParsedFile f;
+            ImportCheckerPass pass;
+
+            for (auto result = fileq->try_pop(f); !result.done(); result = fileq->try_pop(f)) {
+                if (f.file.data(gs).isPackage()) {
+                    auto pkgName = gs.packageDB().getPackageNameForFile(f.file);
+                    if (pkgName.exists()) {
+                        core::Context ctx{gs, core::Symbols::root(), f.file};
+                        f.tree = ast::TreeMap::apply(ctx, pass, std::move(f.tree));
+                        pass.checkImports(ctx);
+                    }
+                }
+
+                resultq->push(std::move(f), 1);
+            }
+        });
+
+        ast::ParsedFile threadResult;
+        for (auto result = resultq->wait_pop_timed(threadResult, WorkerPool::BLOCK_INTERVAL(), gs.tracer());
+             !result.done();
+             result = resultq->wait_pop_timed(threadResult, WorkerPool::BLOCK_INTERVAL(), gs.tracer())) {
+            files.emplace_back(std::move(threadResult));
+        }
+
+        fast_sort(files, [](const auto &a, const auto &b) -> bool { return a.file < b.file; });
+
+        return files;
+    }
+};
+
+} // namespace
+
+// TODO: validate package imports and give errors for packages that don't exist
```

Comment:
What's this TODO for? Should we record it in the followup doc?

Reply:
I'm pretty sure this TODO isn't relevant anymore. I should have taken a pass over the visibility checker to remove TODOs before marking this ready 😬